### PR TITLE
Comments Query Loop: Add discuss options to the block

### PR DESCRIPTION
**Pull request by cbravobernal**
## Description

In this PR, we are adding the option to sort the comments on the Query Loop by Newer/Older.
A review of the copies is kindly appreciated 😄 .

## How has this been tested?

- Added a Comment Query Loop block.
- Changed settings on editor.
- Checked that everything is working both on editor and frontend.

## Screenshots

Loom Video:
<a href="https://www.loom.com/share/212026499a1147aaaddda5232fc95148">
    <p>Edit Post “Post sample” ‹ gutenberg — WordPress - 20 January 2022 - Watch Video</p>
    
  </a>

## Types of changes

New feature, adds an attribute to Comment Query Loop.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/comments-query-loop/order-controls.js
Comment (ntsekouras) [inline packages/block-library/src/comments-query-loop/order-controls.js]: It doesn't seem that the `ToggleGroupControl` is the right one for this control.. I think a combo box control is a better fit here and will accommodate for more options, if they need to be added.

Comment (cbravobernal) [reply packages/block-library/src/comments-query-loop/order-controls.js]: Thanks for the review!!

 It seems that order options only accepts two values here, a long time ago for now. This is what we have when we check the documentation about the comments arguments on that function.
Screenshot 2021-12-13 at 10 29 22

Maybe the combo box would be a better option for OrderBy?
Screenshot 2021-12-13 at 10 32 02

Comment (ntsekouras) [reply packages/block-library/src/comments-query-loop/order-controls.js]: I think the common case is to combine `order and orderBy` in a single control when we want both, but if we want only a control with two values (asc, desc) here, we might even consider a simple toggle control? --cc @jasmussen

Comment (jasmussen) [reply packages/block-library/src/comments-query-loop/order-controls.js]: Nice PR! Yes, as a rule of thumb, if the text labels wrap inside a segmented control, then that control is not the right choice. I would also lean towards a basic toggle in cases where there are only two options. The combobox/dropdown is good for surfacing longer labels and obviously showing many options.

Is the toggle right for sort order? I'd say that depends on the phrasing and if it feels right. Something like this might work:

```
[on] Sort from newest to oldest
```

The implication would be that toggling it off would sort from _oldest to newest_. I think I'd personally like this as a good starting point, since it's a fairly simple toggle and good baseline. 

We can always upgrade to the dropdown, which could be something like:

```
Sort order:
[ Newest to oldest
  Oldest to newest ]
```

What do you think?

File: packages/block-library/src/comment-template/block.json
Comment (gziolo) [inline packages/block-library/src/comment-template/block.json]: In #37196 we discuss using namespaces for the block context so it's clear that it applies to comments and mitigate the risk of any future name collisions. How about we use here and in the parent block `comments/order` instead?

File: packages/block-library/src/comment-template/index.php
Comment (gziolo) [inline packages/block-library/src/comment-template/index.php]: With the default value set in the Comments Query Loop block, it's going to always use the value from the parent block when it's present. It might be not desired if someone wants to inherit global settings set for a site.

File: packages/block-library/src/comments-query-loop/block.json
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/block.json]: See my other comment, it will always pass the default value down to inner blocks through the block context making it impossible to read the site settings.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/edit.js]: This setting should probably be located outside of the advanced group, in their own group specific for the block.

**Review by gziolo — COMMENTED** [review]
Thank you for starting working on this control. My only concern would be that we use the block toolbar to set the number of items per page, and the sorting order is placed in the sidebar.

File: packages/block-library/src/comment-template/index.php
Comment (cbravobernal) [reply packages/block-library/src/comment-template/index.php]: That's true, I will remove the default one then on block.json 

Should we add then an explanation that will inherit the global settings on the helper?

Comment (gziolo) [reply packages/block-library/src/comment-template/index.php]: Once we land #37196, we will be able to contain all those decisions inside `build_comment_query_vars_from_block` and it should give a good opportunity to document how it all works. The idea is that we would use the same helper in all inner blocks that need to query the list of comments.

Comment (cbravobernal) [reply packages/block-library/src/comment-template/index.php]: I will wait for that PR to be landed in order to update this one with all the comments, and that function 😄 🚀

Comment (gziolo) [reply packages/block-library/src/comment-template/index.php]: CI should turn green in a few minutes. Feel free to land it to unblock this PR.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (cbravobernal) [reply packages/block-library/src/comments-query-loop/edit.js]: Added on the toolbar. But I don't like how it shows. Maybe we could consider moving it to a Settings Inspector Options panel.

File: packages/block-library/src/comment-template/edit.js
Comment (cbravobernal) [inline packages/block-library/src/comment-template/edit.js]: I'm not confident with this solution, as both editor and frontend differ by default.

What is your opinion about this? 

Could we solve it with the `toggle box` of `inherit from settings` proposal?

That way, the order would be:
 - If inherit enabled: settings
 - If inherit disabled:  DESC (newer first) by default like the REST API, otherwise chosen by the user with the toggle.

cc: @gziolo @ntsekouras @jasmussen @michalczaplinski @DAreRodz

Comment (gziolo) [reply packages/block-library/src/comment-template/edit.js]: We should read the default values from the server and use them when the user selects no explicit override. I think @michalczaplinski, at some point, had some code that exposes the site settings through the REST API. Am I correct, or was it something else?

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [reply packages/block-library/src/comments-query-loop/edit.js]: It can be in the sidebar. It just shouldn’t be inside the advanced panel where controls related to tweaking HTML code live.

**Review by gziolo — COMMENTED** [review]
It looks like this PR is almost ready. The two remaining tasks I can see are:
- Reading the default order for comments from the server.
- Deciding whether controls for “Discussion Settings” should be in the block toolbar or instead in the sidebar.

**Comment by cbravobernal** [comment]
This is what we see on the toolbar right now. Is there any way to make it wider without custom CSS?

Screenshot 2021-12-16 at 09 08 51

Thanks!

**Comment by cbravobernal** [comment]
> * Deciding whether controls for “Discussion Settings” should be in the block toolbar or instead in the sidebar.

I guess it will be easier to expand in the sidebar.

File: packages/block-library/src/comment-template/edit.js
Comment (michalczaplinski) [reply packages/block-library/src/comment-template/edit.js]: Yes! I wanted to have `comments_per_page` exposed in the REST API. Here are quite complete instructions on how to do it, they should apply to the `order` property as well: 
https://wordpress.slack.com/archives/C02QB2JS7/p1635376326175900

For those without access to the WP slack, here's how it can be done. It should be added to `/lib/rest-api.php`
```php
function register_block_core_comment_template_order_setting() {
  register_setting(
	  'discussion',
	  'order',
	  array(
		  'show_in_rest' => true,
	  )
  );
}

add_action( 'rest_api_init', 'register_block_core_comment_template_order_setting', 10 );
```

Comment (cbravobernal) [reply packages/block-library/src/comment-template/edit.js]: Wow, this is a really quick way to fix it, it will save me a lot of time 🎉 

Would it be a good idea to create an endpoint to retrieve all the discussion settings? We could use `register_rest_route`, get all Discussion settings and use them for all the block rendering.

Comment (gziolo) [reply packages/block-library/src/comment-template/edit.js]: I don't think you need to create a new endpoint for that. As far as I'm aware the code that @michalczaplinski shared, adds the setting to the endpoint that is already in use in Gutenberg.

Comment (gziolo) [reply packages/block-library/src/comment-template/edit.js]: I have an update in regards to these settings. I talked to @spacedmonkey and he explained that the https://make.wordpress.org/core/wp-json/wp/v2/settings endpoint is only available for admins. There is a long-standing Tract ticket that we would need to land: https://core.trac.wordpress.org/ticket/48885 – the idea is to expose site settings in the read-only mode for all user roles.

As a short-term solution, we could always expose those setting using the `block_editor_settings_all` filter. See am example of how to share the settings from the server:

https://github.com/WordPress/gutenberg/blob/a92586c6ba83e735e102867bc69584b6af856643/lib/compat/wordpress-5.9/default-editor-styles.php#L17-L41

**Comment by cbravobernal** [comment]
On the sidebar is easier to handle.

Screenshot 2021-12-20 at 17 51 07

The code is working, but the editor comments won't get updated if the user is not an admin.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (ntsekouras) [inline packages/block-library/src/comments-query-loop/edit.js]: We need to wrap this in a `useEffect` because it causes side effects.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (ntsekouras) [inline packages/block-library/src/comments-query-loop/edit.js]: Shouldn't we also handle the default `perPage` assignment in `Comments Query` and not in the `Comments Template` block as [is done now](https://github.com/WordPress/gutenberg/pull/37297/files#diff-885985bee90d7d51f079b99d59d8562cba40b4a6f64258e86fc81b0dc0f11b87R150)? This way it will always be filled by `context` and the errors in the `Comments Query` controls will go away.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [reply packages/block-library/src/comments-query-loop/edit.js]: Actually, I think that we should remove this code because we don't want to save the current default order as an attribute for the block. We should only use the value in the UI when the user wants to change the default value – a sort of fallback or starting point depending on how you think about it.

In the current version, this would also create an undo level for the attribute change.

File: packages/block-library/src/comments-query-loop/toolbar.js
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/toolbar.js]: Would it work if we flip the check?
```suggestion
									order: order !== 'asc' ? 'desc" : 'asc',
```

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [reply packages/block-library/src/comments-query-loop/edit.js]: If the values are not set for `perPage`, and/or `order` we want them to be read from the site settings. Otherwise, when we save explicitly the current site settings in the serialized block, any further global change won't be reflected in the block. I guess it's up to debate which approach is better because both have pros and cons, but at least that was my thinking when I worked on the initial implementation for `perPage`.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (ntsekouras) [inline packages/block-library/src/comments-query-loop/edit.js]: Since you've created this new file `edit/toolbar` we should remove the old one in the root folder, no?

File: packages/block-library/src/comments-query-loop/edit/toolbar.js
Comment (ntsekouras) [inline packages/block-library/src/comments-query-loop/edit/toolbar.js]: >This is what we see on the toolbar right now. Is there any way to make it wider without custom CSS?

You'll have to change this value here.

Also in order to have a wider popover you can add an extra css class. At least this is what we're doing for `Query Loop`: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/edit/query-toolbar.js#L47

File: packages/block-library/src/comments-query-loop/edit.js
Comment (cbravobernal) [reply packages/block-library/src/comments-query-loop/edit.js]: The reason for having this is to show a default position for the toggle, depending on what is defined on the settings version. Right now is not working if you are not an admin.

Maybe we could add an option of "Inherit from settings", default enabled. And, in case you disable it, we can set a default order attribute.

Meanwhile, I updated the code in order not to cause side effects or mess with undo buttons.

**Comment by cbravobernal** [comment]
I still have some issues in case anyone wants to keep working on this issue while I'm on Christmas Holidays:

- [x] Setting the attributes as null will cause that the render will differ from default settings. This is because REST API has different default values than discussion settings. Maybe is better to get the values always from the settings on the block rendering instead of using the REST API.
- [x] The frontend version also will be different than the block version if the `default_comments_page` setting is different than the default one.

**Comment by cbravobernal** [comment]
Hi again!
I updated the code in order to have an inherit option. Feel free to test and share feedback!

<a href="https://www.loom.com/share/b5c8a9c1f1294572a553d55993c228d7"> <p>Loom Message - 13 January 2022 - Watch Video</p>  </a>

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/edit.js]: For consistency, `BlockControls` wrapper could be moved to the `CommentsToolbar` components. More importantly, do we need the same controls in the toolbar in the sidebar?

File: lib/compat/experimental/blocks.php
Comment (gziolo) [inline lib/compat/experimental/blocks.php]: What's the reason that the logic related to `order` is split into two chunks? Can we consolidate it in one place? When the context is set, we don't need to call `get_option` at all. Unless I'm missing something.

Comment (gziolo) [reply lib/compat/experimental/blocks.php]: In my testing, when I have the `inherit` attribute on the Comments Query Loop block enabled, the context value still gets processed here. It means that either the attribute should be set to `null` in the wrapping block or the check should take `inherit` into account.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/edit.js]: I'm not convinced we really need those side effects here. When the `inherit` value is set to `false` (default) then we can read values directly from settings in the components that render controls. In case the user toggles the `inherit` control then we can set the new value for `inherit`. In addition, we can use the same `setAttributes` call to  set values for `order` and `perPage` to `null`, when toggling to `true`, and to values read from settings, when toggling to `false`.

If we set `order` and `perPage` to `null` when enabling discussion setting inheriting then we should have the issue on the server resolved when the last stored values are read.

**Review by gziolo — COMMENTED** [review]
I did another pass on the code reviews and testing and I reported my findings. We need to look closer at the logic around the new "inherit" toggle but otherwise, everything looks good.

File: packages/block-library/src/comments-query-loop/edit.js
Comment (cbravobernal) [reply packages/block-library/src/comments-query-loop/edit.js]: We don't need them, I will remove the toolbar, if we check the  Query Loop block, the Toolbar only contains the posts per page.

In this case, we could move all of them to the Sidebar, as we will need the inherit option on both sides if we want to have different attributes both in the toolbar and sidebar, causing difficulties for the users.

Also, I would like to update the sort option to a selector, as we have on Query Loop block also. Will update the code and let's see if it is better!

**Comment by cbravobernal** [comment]
Video updated 😄 

<a href="https://www.loom.com/share/212026499a1147aaaddda5232fc95148">
    <p>Edit Post “Post sample” ‹ gutenberg — WordPress - 20 January 2022 - Watch Video</p>
    
  </a>

File: packages/block-library/src/comments-query-loop/edit.js
Comment (gziolo) [inline packages/block-library/src/comments-query-loop/edit.js]: I recommend returning only `__experimentalDiscussionSettings` here, because otherwise, this component will re-render every time any setting changes.

File: lib/compat/experimental/blocks.php
Comment (gziolo) [inline lib/compat/experimental/blocks.php]: Nit: If we assign the initial value to `$comment_args['order']` instead of `$order` we should not need this part after `else`.

**Comment by gziolo** [comment]
I need to test everything again, but it feels like everything is ready. I left two small nitpicks that don't impact the functionality.

File: packages/block-library/src/comment-template/edit.js
Comment (gziolo) [inline packages/block-library/src/comment-template/edit.js]: Similar as in other place:
```suggestion
	const { commentOrder, commentsPerPage } = useSelect( ( select ) => {
		const { getSettings } = select( blockEditorStore );
		return getSettings().__experimentalDiscussionSettings;
	} );
```

**Review by gziolo — APPROVED** [review]
Let's ship this version. I think the way it works now is pretty good. Everything works correctly in the case settings are inherited from the site settings and when some custom values get selected. Toggling settings inheriting also works great.

I'm testing using the legacy Post Comments block next to the Comments Query Loop block added to the Single Post template. I noticed that the order of pages differs between those two. I bet we need some more legacy logic related to the mix of first/last page and older/newer settings:

Screenshot 2022-01-25 at 16 03 09

Let's look into it the follow-up PR.

File: packages/block-library/src/comments-query-loop/editor.scss
Comment (ockham) [inline packages/block-library/src/comments-query-loop/editor.scss:3]: While working on #40522, @DAreRodz, @michalczaplinski and I found that this file doesn't seem to be used.

It seems that it was introduced to fix the issue mentioned [here](https://github.com/WordPress/gutenberg/pull/37297#issuecomment-995535517), but it appears the relevant settings were [later moved to the inspector controls](https://github.com/WordPress/gutenberg/pull/37297#issuecomment-998101414) in the sidebar, making the styling obsolete.

Just documenting this here for posterity 😄

**Pull request merged**
